src/transformStateWithClones.js

```diff
@@ -3,11 +3,36 @@
 /**
  * @param {Object} state
  * @param {Object[]} actions
- *
+ *git
  * @return {Object[]}
  */
+function deepClone(obj) {
+  return JSON.parse(JSON.stringify(obj));
+}
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = deepClone(state);
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    switch (obj.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, obj.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of obj.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      case 'clear':
+        stateCopy = {};
+        break;
+    }
+
```

Comment:
[CODE STYLE 3]: switch/case should always have default case for error handling.

You have forgotten to add default case with Error handling. Just add to default case `throw Error`